### ISSUE TITLE
Bizarre behavior after updating from librdkafka v0.11.5 to v1.3.0

### ISSUE THREAD

**Issue by nekuwafers** [edenhill/librdkafka#2820]
Description
===========
Hey edenhill, after updating kafka from v0.11.5 to v1.3.0, our implementation has been seeing some bizarre issues. When we create a KafkaProducer, we receive a seg fault, and right before we do we see an odd log:

[thrd:app]: Failed to initialize rdkafka#producer-1: 1 background thread(s) did not initialize within 60 seconds

After searching through the librdkafka library, we believe that log is being printed here: https://github.com/edenhill/librdkafka/blob/master/src/rdkafka.c#L2364, which belongs in a block of code that should be unreachable. We did a bit of debugging and found that, if we create our producer with a nullptr as the globalConfig, everything works fine. However, if we pass in a globalConfig, even if it doesn't have anything configured, we see this issue. Additionally, we tried printing out the RdKafka version using RdKafka::version_str() before we create our producer, just as a sanity test, and this resolved the issue. If we printed anything else or waited for a few seconds, the issue wasn't resolved, but calling RdKafka::version_str() fixed it (we didn't have to print it out, just calling RdKafka::version_str() did the trick). This definitely doesn't feel like a very stable solution, so we were wondering if you had any idea what was going on here, or were aware of a better solution to the issue.

Checklist
=========

 - [x] librdkafka version: 1.3.0
 - [x] librdkafka client configuration:  auto.create.topics.enable: "false", log.flush.scheduler.interval.ms: "30000", offsets.topic.num.partitions: "12", offsets.topic.segment.bytes: "10485760", offsets.retention.minutes: "60", log.index.size.max.bytes: "1048576", log.retention.bytes: "1310720", log.segment.bytes: "655360"
 - [x] Operating system: Centos 7.5.1804 (x64)
 - [ ] Provide logs (with `debug=..` as necessary) from librdkafka
 - [ ] Provide broker log excerpts
 - [ ] Critical issue

**Comment by edenhill** [comment]
This sounds like memory corruption.

Can you try running your application with valgrind and see if it reports something useful?

**Comment by nekuwafers** [comment]
Nothing shows up on valgrind (like it's literally blank)

![image](https://user-images.githubusercontent.com/43799667/79241499-bf3dfb00-7e38-11ea-8b46-5f427c7c6b68.png)

The valgrind command run was: valgrind --log-file=out.log --leak-check=full --show-leak-kinds=all ./Executable

**Comment by edenhill** [comment]
So it doesn't crash with valgrind?
Does the program work  when running under valgrind or does it just hang?

The next alternative is to fire up good old gdb and inspect the stack frames.

**Comment by nekuwafers** [comment]
It still seg faults with valgrind. However, when I added the --trace-children=yes flag, there was quite a bit of output, a lot of which pointed to mallocs made by trying to create the kafka producer (it should be noted that the seg fault in question happens because when we try to create a producer, it returns a nullptr because the background threads couldn't be initialized, and when this is passed into the creation of the Topic, we get a seg fault from bad access).

I looked through the logs from the --trace-children=yes, and i didn't find anything too striking (though I'm not exactly a valgrind expert). I did find quite a few block losses of this format though:

![image](https://user-images.githubusercontent.com/43799667/79268319-33d86000-7e60-11ea-9377-b3bdc0f37fec.png)

I'm not exactly sure what libcrypto is, but does this look like anything to you?

**Comment by edenhill** [comment]
Are you forking your process?
If so, make sure to create your clients in the same process as they will be used as the internal librdkafka threads will not be available in the fork.
I.e., you can't create a client in the parent process and then use it in a forked child process, the client must be created in the child.

**Comment by edenhill** [comment]
The libcrypto leaks are expected (there are valgrind suppressions available here https://github.com/edenhill/librdkafka/blob/master/tests/librdkafka.suppressions), it is once-per-runtime allocations of OpenSSL, such as for error strings, etc. They can be safely ignored.

**Comment by nekuwafers** [comment]
Ok so I think I figured out what's up. We previously created our producer in an  __attribute__((constructor)) method (so it was created when the library was loaded). Your comment about forked processes got me thinking that maybe that was the issue, so I got rid of that and then created the producer at a point in our actual executed code, and it looks like that resolved the issue! Not sure why it worked in kafka 0.11.5 and not 1.3.0, but hey, all's well that ends well.
